README.md
[![Travis Status](https://travis-ci.org/n3okill/enfsaddins.svg)](https://travis-ci.org/n3okill/enfsaddins)
[![Appveyor status](https://ci.appveyor.com/api/projects/status/ng4k20sk93vm04o1?svg=true)](https://ci.appveyor.com/project/n3okill/enfsaddins)
[![Codacy Badge](https://api.codacy.com/project/badge/grade/f564e01b160f403e83bf8d63f50e280a)](https://www.codacy.com/app/n3okill/enfsaddins)
[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_SM.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=64PYTCDH5UNZ6)

[![NPM](https://nodei.co/npm/enfsaddins.png)](https://nodei.co/npm/enfsaddins/)

enfsaddins
=========
Additional methods for node fs module

**enfs** stands for [E]asy [N]ode [fs]

This module is intended to work as a sub-module of [enfspatch](https://www.npmjs.com/package/enfspatch)

Description
-----------
This module will change some behaviors of fs module from node
such as creating a queue for opening files when limit is reached,
catching the error's and proceeding with the process when possible.
- This module will change the behaviour of:
  * exists
  * existsSync
  
Usage
-----
`enfsaddins`


Valid usage but not the better:

```js
    var fs = require("fs");
    var enfs = require("enfsaddins")(fs);
```

It's better to use

```js
    var enfs = require("enfspatch");
```

this will use [enfspatch](https://www.npmjs.com/package/enfspatch) module that implements this module
and allows access to all it's methods



Errors
------
All the methods follows the node culture.
- Async: Every async method returns an Error in the first callback parameter
- Sync: Every sync method throws an Error.


Additional Methods
------------------
- [exists](#exists)
- [existsSync](#exists)
- [existAccess](#existaccess)
- [existAccessSync](#existaccess)
- [existStat](#existstat)
- [existLStat](#existlstat)
- [existFStat](#existfstat)
- [existStatSync](#existstat)
- [existLStatSync](#existlstat)
- [existFStatSync](#existfstat)
- [existStatIsDirectory](#existstatisdirectory)
- [existLStatIsDirectory](#existstatisdirectory)
- [existFStatIsDirectory](#existstatisdirectory)
- [existStatIsDirectorySync](#existstatisdirectory)
- [existLStatIsDirectorySync](#existstatisdirectory)
- [existFStatIsDirectorySync](#existstatisdirectory)
- [existStatIsFile](#existstatisfile)
- [existLStatIsFile](#existstatisfile)
- [existFStatIsFile](#existstatisfile)
- [existStatIsFileSync](#existstatisfile)
- [existLStatIsFileSync](#existstatisfile)
- [existFStatIsFileSync](#existstatisfile)
- [existIsSymlink](#existissymlink)
- [existIsSymlinkSync](#existissymlink)


### exists
  - **exists(path, callback)**

> Change the natural behaviour of fs.exists to the node culture, it will return an error 
in the first callback parameter.
As exists is deprecated if it cease to exist then exists will use (#existStat) instead

Sync: 
  - **existsSync(path)**

```js
    enfs.exists("/etc/passwd", function(err, itemExists){
        console.log(itemExists ? 'it\'s there' : 'no passwd!');
    });
```
check: [fs.exists](https://nodejs.org/api/fs.html#fs_fs_exists_path_callback)

### existAccess
  - **existAccess(path, [mode], callback)**

> Will use fs.access to check if the item exists in the file system and if the process
as access to the item.

Sync: 
  - **existAccessSync(path, [mode])**


```js
    enfs.existAccess("/etc/passwd", function(err,itemExists){
        console.log(itemExists ? "it\'s there and have access" : "don\'t exist or don\'t have access");
    });
```
check: [fs.access](https://nodejs.org/api/fs.html#fs_fs_access_path_mode_callback)

### existStat 
### existLStat
- **existStat(path,callback)**
- **existLStat(path,callback)**

> Will use fs.stat to check if the item exists in the file system.

Sync: 
  - **existStatSync(path)**
  - **existLStatSync(path)**

```js
    enfs.existStat("/etc/passwd", function(err,itemExists){
        console.log(itemExists ? "it\'s there" : "don\'t exist");
    });
```
check: [fs.stat](https://nodejs.org/api/fs.html#fs_fs_stat_path_callback)

```js
    enfs.existLStat("/etc/passwd", function(err,itemExists){
        console.log(itemExists ? "it\'s there" : "don\'t exist");
    });
```
check: [fs.lstat](https://nodejs.org/api/fs.html#fs_fs_lstat_path_callback)

### existFStat
  - **existFStat(fd,callback)**

> Will use fs.fstat to check if the item exists in the file system.

Sync: 
  - **existFStatSync(path)**

```js
    enfs.existFStat(enfs.openSync("/etc/passwd","r"), function(err,itemExists){
        console.log(itemExists ? "it\'s there" : "don\'t exist");
    });
```
check: [fs.fstat](https://nodejs.org/api/fs.html#fs_fs_fstat_fd_callback)


### existStatIsDirectory 
### existLStatIsDirectory
### existFStatIsDirectory
- **existStatIsDirectory(path,callback)**
- **existLStatIsDirectory(path,callback)**
- **existFStatIsDirectory(path,callback)**

> Will use fs.stat or fs.lstat or fs.fstat to check if the item exists in the file system,
and if it's a directory.
This method is just a shortcut to check if an item exists in the file system and it's type

Sync: 
  - **existStatIsDirectorySync(path)**
  - **existLStatIsDirectorySync(path)**
  - **existFStatIsDirectorySync(path)**

```js
    enfs.existStatIsDirectory("/etc", function(err,isDirectory,stat){
        console.log(isDirectory ? "it's a directory" : "don\'t exist or it's not a directory.");
    });
```
check: 
[fs.stat](https://nodejs.org/api/fs.html#fs_fs_stat_path_callback)
[fs.lstat](https://nodejs.org/api/fs.html#fs_fs_lstat_path_callback)
[fs.fstat](https://nodejs.org/api/fs.html#fs_fs_fstat_fd_callback)


### existStatIsFile 
### existLStatIsFile
### existFStatIsFile
- **existStatIsFile(path,callback)**
- **existLStatIsFile(path,callback)**
- **existFStatIsFile(path,callback)**

> Will use fs.stat or fs.lstat or fs.fstat to check if the item exists in the file system,
and if it's a file.
This method is just a shortcut to check if an item exists in the file system and it's type

Sync: 
  - **existStatIsFileSync(path)**
  - **existLStatIsFileSync(path)**
  - **existFStatIsFileSync(path)**

```js
    enfs.existStatIsFile("/etc/passwd", function(err,isFile,stat){
        console.log(isFile ? "it's a file" : "don\'t exist or it's not a file.");
    });
```
check: 
[fs.stat](https://nodejs.org/api/fs.html#fs_fs_stat_path_callback)
[fs.lstat](https://nodejs.org/api/fs.html#fs_fs_lstat_path_callback)
[fs.fstat](https://nodejs.org/api/fs.html#fs_fs_fstat_fd_callback)


### existIsSymlink 
- **existIsSymlink(path,callback)**

> Will use fs.lstat to check if the item exists in the file system,
and if it's a symbolic link.
This method is just a shortcut to check if an item exists in the file system and it's type

Sync: 
  - **existIsSymlinkSync(path)**

```js
    enfs.existIsSymlink("/etc/symlink", function(err,isSymlink,stat){
        console.log(isSymlink ? "it's a symlink" : "don\'t exist or it's not a symlink.");
    });
```
check: 
[fs.lstat](https://nodejs.org/api/fs.html#fs_fs_lstat_path_callback)




License
-------

Creative Commons Attribution 4.0 International License

Copyright (c) 2016 Joao Parreira <joaofrparreira@gmail.com> [GitHub](https://github.com/n3okill)

This work is licensed under the Creative Commons Attribution 4.0 International License. 
To view a copy of this license, visit [CCA4](http://creativecommons.org/licenses/by/4.0/).




CHANGELOG.md
1.0.0 / 2017-02-14
==================
  * general code improvement
  * dev dependencies upgrade
  * major version

0.0.3 / 2016-04-07
==================
  * es6 code update
  * dependencies update

0.0.2 / 2016-02-23
==================
  * remove enfspatch from devDependencies
  
0.0.1 / 2016-02-23
==================
  * first release


test/addins.js
/**
 * @project enfsaddins
 * @filename addins.js
 * @description tests for enfsaddins
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.1.1
 */

/* global describe, require, it, __filename */

"use strict";

//avoid messing with original fs module
const copyFs = require("clone")(require("fs"));

describe("enfsaddins", function () {
    const fs = require("../")(copyFs);
    describe("> existStat", function () {
        describe("> async", function () {
            function existStat(path, result, done) {
                fs.existStat(path, function (err, res) {
                    (err === null).should.be.equal(true);
                    res.should.be.equal(result);
                    done();
                });
            }

            it("should test true for async __filename", function (done) {
                existStat(__filename, true, done);
            });
            it("should test false for async non-existent file", function (done) {
                existStat("/not/existent/file", false, done);
            });
        });
        describe("> sync", function () {
            it("should test true for sync __filename", function () {
                fs.existStatSync(__filename).should.be.equal(true);
            });
            it("should test false for sync non-existent file", function () {
                fs.existStatSync("/not/existent/file").should.be.equal(false);
            });
        });
    });
    describe("> existAccess", function () {
        describe("> async", function () {
            function existAccess(path, result, done) {
                fs.existAccess(path, function (err, res) {
                    (err === null).should.be.equal(true);
                    res.should.be.equal(result);
                    done();
                });
            }

            it("should test true for async __filename", function (done) {
                existAccess(__filename, true, done);
            });
            it("should test false for async non-existent file", function (done) {
                existAccess("/not/existent/file", false, done);
            });
        });
        describe("> sync", function () {
            it("should test true for sync __filename", function () {
                fs.existAccessSync(__filename).should.be.equal(true);
            });
            it("should test false for sync non-existent file", function () {
                fs.existAccessSync("/not/existent/file").should.be.equal(false);
            });
        });
    });
    describe("> multiple methods of exist and is of type", function () {
        describe("> async", function () {
            const methods = ["existStatIsDirectory", "existLStatIsDirectory", "existFStatIsDirectory",
                "existStatIsFile", "existLStatIsFile", "existFStatIsFile",
                "existIsSymlink"];
            it("should test methods existence", function () {
                methods.forEach(function (name) {
                    (typeof fs[name]).should.be.equal("function");
                });
            });
            it("should test stat isDirectory method and it's result", function (done) {
                fs.existStatIsDirectory(__dirname, function (err, result) {
                    (err === null).should.be.equal(true);
                    result.should.be.equal(true);
                    fs.existStatIsDirectory(__filename, function (errFile, resultFile) {
                        (errFile === null).should.be.equal(true);
                        resultFile.should.be.equal(false);
                        done();
                    });
                });
            });
            it("should test stat isFile method and it's result", function (done) {
                fs.existStatIsFile(__filename, function (err, result) {
                    (err === null).should.be.equal(true);
                    result.should.be.equal(true);
                    fs.existStatIsFile(__dirname, function (errDir, resultDir) {
                        (errDir === null).should.be.equal(true);
                        resultDir.should.be.equal(false);
                        done();
                    });
                });
            });
        });
        describe("> sync", function () {
            it("should test methods existence", function () {
                const methods = ["existStatIsDirectorySync", "existLStatIsDirectorySync", "existFStatIsDirectorySync",
                    "existStatIsFileSync", "existLStatIsFileSync", "existFStatIsFileSync",
                    "existIsSymlinkSync"];
                methods.forEach(function (item) {
                    (typeof fs[item]).should.be.equal("function");
                });
            });
            it("should test stat isDirectory method and it's result", function () {
                fs.existStatIsDirectorySync(__dirname).should.be.equal(true);
                fs.existStatIsDirectorySync(__filename).should.be.equal(false);
            });
            it("should test stat isFile method and it's result", function () {
                fs.existStatIsFileSync(__filename).should.be.equal(true);
                fs.existStatIsFileSync(__dirname).should.be.equal(false);
            });
        });
    });
});


lib/addins.js
/**
 * @project enfsaddins
 * @filename addins.js
 * @description Add functions to the fs module
 * @author Joao Parreira <joaofrparreira@gmail.com>
 * @copyright Copyright(c) 2016 Joao Parreira <joaofrparreira@gmail.com>
 * @licence Creative Commons Attribution 4.0 International License
 * @createdAt Created at 18-02-2016.
 * @version 0.1.1
 */

"use strict";

function noop() {
}

function addins(fs) {
    const fnsExist = [["existStat", "stat"], ["existLStat", "lstat"], ["existFStat", "fstat"]];
    fnsExist.forEach((fnE) => {
        fs[fnE[0]] = function(arg, callback) { callback && fs[fnE[1]](arg, (err, stat) => callback(null, !err, err || stat)); }
    });

    const fnExistSync = [["existStatSync", "statSync"], ["existLStatSync", "lstatSync"], ["existFStatSync", "fstatSync"]];
    fnExistSync.forEach((fnE) => {
        fs[fnE[0]] = function(arg) {
            try {
                fs[fnE[1]](arg);
                return true;
            } catch (err) {
                return false;
            }
        }
    });

    fs.existAccess = function (path, mode, callback) {
        callback = callback || noop;
        if (typeof mode === "function") {
            callback = mode;
            mode = null;
        }
        fs.access(path, mode, (err) => {
            callback(null, !err);
        });
    };
    fs.existAccessSync = function (path, mode) {
        try {
            fs.accessSync(path, mode);
            return true;
        } catch (err) {
            return false;
        }
    };
    fs.exists = ((function (exists) {
        return function (path, callback) {
            callback = callback || noop;
            if (exists) {
                exists(path, (e) => {
                    callback(null, e);
                });
            } else {
                fs.existStat(path, callback);
            }
        };
    })(fs.exists));

    fs.existsSync = fs.existsSync || fs.existStatSync;

    const fnsAsync = [
        {name: "existStatIsDirectory", fsFn: "stat", statType: "isDirectory"},
        {name: "existLStatIsDirectory", fsFn: "lstat", statType: "isDirectory"},
        {name: "existFStatIsDirectory", fsFn: "fstat", statType: "isDirectory"},
        {name: "existStatIsFile", fsFn: "stat", statType: "isFile"},
        {name: "existLStatIsFile", fsFn: "lstat", statType: "isFile"},
        {name: "existFStatIsFile", fsFn: "fstat", statType: "isFile"},
        {name: "existIsSymlink", fsFn: "lstat", statType: "isSymbolicLink"}
    ];
    const fnsSync = [
        {name: "existStatIsDirectorySync", fsFn: "statSync", statType: "isDirectory"},
        {name: "existLStatIsDirectorySync", fsFn: "lstatSync", statType: "isDirectory"},
        {name: "existFStatIsDirectorySync", fsFn: "fstatSync", statType: "isDirectory"},
        {name: "existStatIsFileSync", fsFn: "statSync", statType: "isFile"},
        {name: "existLStatIsFileSync", fsFn: "lstatSync", statType: "isFile"},
        {name: "existFStatIsFileSync", fsFn: "fstatSync", statType: "isFile"},
        {name: "existIsSymlinkSync", fsFn: "lstatSync", statType: "isSymbolicLink"}
    ];

    fnsAsync.forEach((fn) => {
        fs[fn.name] = function(path, callback) { callback && fs[fn.fsFn](path, (err, stat) => callback(null, err ? false : !!stat[fn.statType](), err || stat)); }
    });

    fnsSync.forEach((fn) => {
        fs[fn.name] = function(path) {
            try {
                const stat = fs[fn.fsFn](path);
                return !!stat[fn.statType]();
            } catch (err) {
                return false;
            }
        };
    });
    return fs;
}


module.exports = addins;